main.tf

```diff
@@ -1,26 +1,53 @@
 locals {
-  enabled    = "${var.enabled == "true" ? true : false }"
-  id         = "${local.enabled ? lower(join(var.delimiter, compact(concat(list(var.namespace, var.stage, var.name), var.attributes)))) : ""}"
-  name       = "${local.enabled ? lower(format("%v", var.name)) : ""}"
-  namespace  = "${local.enabled ? lower(format("%v", var.namespace)) : ""}"
-  stage      = "${local.enabled ? lower(format("%v", var.stage)) : ""}"
-  attributes = "${local.enabled ? lower(format("%v", join(var.delimiter, compact(var.attributes)))) : ""}"
+  enabled = "${var.enabled == "true" ? true : false }"
 
-  tags = "${
-      merge( 
-        map(
-          "Name", "${local.id}",
-          "Namespace", "${local.namespace}",
-          "Stage", "${local.stage}"
-        ), var.tags
-      )
-    }"
+  id = "${lower(join(local.delimiter, compact(concat(list(local.namespace, local.stage, local.name), local.attributes))))}"
 
+  # selected_name : Select which value to use, the one from context, or the one from the var
+  # name: Remove spaces, make lowercase
+
+  selected_name       = ["${compact(concat(list(var.name), local.context_local["name"]))}"]
+  name                = "${lower(join("", split(" ", local.selected_name[0])))}"
+  selected_namespace  = ["${compact(concat(local.context_local["namespace"], list(var.namespace)))}"]
+  namespace           = "${lower(join("", split(" ", local.selected_namespace[0])))}"
+  selected_stage      = ["${compact(concat(local.context_local["stage"], list(var.stage)))}"]
+  stage               = "${lower(join("", split(" ", local.selected_stage[0])))}"
+  selected_attributes = ["${distinct(compact(concat(var.attributes, local.context_local["attributes"])))}"]
+  attributes          = "${split("~^~", lower(join("~^~", local.selected_attributes)))}"
+  selected_delimiter  = ["${distinct(compact(concat(local.context_local["delimiter"], list(var.delimiter))))}"]
+  delimiter           = "${join("", split(" ", local.selected_delimiter[0]))}"
+  # Merge the map of empty values, with the variable context, so that context_local always contains all map keys
+  context_local = "${merge(local.context_context, var.context)}"
+  # Only maps that contain all the same attribute types can be merged, so they have been set to list
+  context_context = {
```

Comment:
can we rename this? maybe call it `null_context`?

Reply:
I'm renaming it to `context_struct`, since it is the structure of context with all variables.